ntoskrnl/se/debug.c

```diff
@@ -0,0 +1,388 @@
+    {
+        DbgPrint("SE_RM_CONTROL_VALID\n");
+    }
+
+    if (SdControl & SE_SELF_RELATIVE)
+    {
+        DbgPrint("SE_SELF_RELATIVE\n");
+    }
+}
+
+/**
+ * @brief
+ * Dumps each security identifier (SID) of an access token to debugger.
+ */
+static
+VOID
+SepDumpSidsOfToken(
+    _In_ PSID_AND_ATTRIBUTES Sids,
+    _In_ ULONG SidCount)
+{
+    ULONG SidIndex;
+    ULONG SidNumber;
+    UNICODE_STRING SidString;
+
+    /* Loop all SIDs and dump them */
+    SidNumber = 1;
+    for (SidIndex = 0; SidIndex < SidCount; SidIndex++)
+    {
+        RtlConvertSidToUnicodeString(&SidString, Sids[SidIndex].Sid, TRUE);
+        DbgPrint("%lu# %wZ\n", SidNumber, &SidString);
+        RtlFreeUnicodeString(&SidString);
+
+        /* Increment the SID number as we go further */
+        SidNumber++;
+    }
+}
+
+/* PUBLIC FUNCTIONS ***********************************************************/
+
+/**
+ * @brief
+ * Dumps debug information of a security descriptor to the debugger.
+ */
+VOID
+SepDumpSdDebugInfo(
+    _In_ PISECURITY_DESCRIPTOR SecurityDescriptor)
+{
+    UNICODE_STRING SidString;
+    PSID OwnerSid, GroupSid;
+    PACL Dacl, Sacl;
+
+    /* Cache the necessary security buffers to dump info from */
+    OwnerSid = SepGetOwnerFromDescriptor(SecurityDescriptor);
+    GroupSid = SepGetGroupFromDescriptor(SecurityDescriptor);
+    Sacl = SepGetSaclFromDescriptor(SecurityDescriptor);
+    Dacl = SepGetDaclFromDescriptor(SecurityDescriptor);
+
+    DbgPrint("================== SECURITY DESCRIPTOR DUMP INFO ==================\n");
+    DbgPrint("SD pointer -> %p\n", SecurityDescriptor);
+    DbgPrint("SD Revision -> %u\n", SecurityDescriptor->Revision);
+    DbgPrint("SD Control:\n");
+    SepDumpSdControlInfo(SecurityDescriptor->Control);
+
+    /* Dump the Owner SID if the SD belongs to an owner */
+    if (OwnerSid)
+    {
+        RtlConvertSidToUnicodeString(&SidString, OwnerSid, TRUE);
+        DbgPrint("SD Owner SID -> %wZ\n", &SidString);
+        RtlFreeUnicodeString(&SidString);
+    }
+
+    /* Dump the Group SID if the SD belongs to a group */
+    if (GroupSid)
+    {
+        RtlConvertSidToUnicodeString(&SidString, GroupSid, TRUE);
+        DbgPrint("SD Group SID -> %wZ\n", &SidString);
+        RtlFreeUnicodeString(&SidString);
+    }
+
+    /* Dump the ACL contents of SACL if this SD has one */
+    if (Sacl)
+    {
+        SepDumpAclInfo(Sacl, TRUE);
+    }
+
+    /* Dump the ACL contents of DACL if this SD has one */
+    if (Dacl)
+    {
+        SepDumpAclInfo(Dacl, FALSE);
+    }
+}
+
+/**
+ * @brief
+ * Dumps debug information of an access token to the debugger.
+ */
+VOID
+SepDumpTokenDebugInfo(
+    _In_ PTOKEN Token)
+{
+    UNICODE_STRING SidString;
+
+    /* Dump relevant token info */
+    DbgPrint("================== ACCESS TOKEN DUMP INFO ==================\n");
+    DbgPrint("Token pointer -> %p\n", Token);
+    DbgPrint("Token source -> %s\n", Token->TokenSource.SourceName);
```

Comment:
Please test this, this should improve the output
(In your original example:
`(ntoskrnl\se\debug.c:438) Token source -> User32  9&`
some extra garbage was showing up because the SourceName is a counted, not NULL-terminated buffer)
```suggestion
    DbgPrint("Token source -> \"%-.*s\"\n",
             RTL_NUMBER_OF(Token->TokenSource.SourceName),
             Token->TokenSource.SourceName);
```